src/engines/pallet_engine/steps.rs

```diff
@@ -0,0 +1,135 @@
+#[derive(Debug)]
+pub(super) enum Steps {
+	/// Import statements for pallet
+	RuntimePalletImport((TokenStream2, usize)),
+	/// Every pallet must impl pallet::Config for Runtime
+	RuntimePalletConfiguration((TokenStream2, usize)),
+	/// The runtime/Cargo.toml needs an import for the pallet being inserted
+	/// This includes features [try-runtime, runtime-benchmarks, std], path information for `git` or local path
+	RuntimePalletDependency(Dependency),
+	/// ConstructRuntime! entry
+	ConstructRuntimeEntry(AddPalletEntry),
+	/// Include a `list_benchmarks!` entry
+	ListBenchmarks(String),
+	/// Does pallet require a genesis configuration?
+	ChainspecGenesisConfig(String),
+	/// ChainSpec imports if necessary
+	ChainspecGenesisImport((TokenStream2, usize)),
+	/// Node specific imports if the above two are required
+	NodePalletDependency(Dependency),
+	/// PalletEngine State transitions
+	SwitchTo(State),
+}
+/// Some rules to follow when constructing steps:
+/// The pallet engine state expects to go as edits would, i.e. top to bottom lexically
+/// So it makes sense for any given file, to first include an import, then items that refer to it
+/// In case of a pallet, you'd always put `RuntimePalletImport`, `RuntimePalletConfiguration`, `ConstructRuntimeEntry` in that order.
+pub(super) fn step_builder(
+	pallet: AddPallet,
+	&TomlEditor { ref workspace, .. }: &TomlEditor,
+) -> Result<Vec<Steps>> {
+	let mut steps: Vec<Steps> = vec![];
+	match pallet {
+		// Adding a pallet-parachain-template requires 5 distinct steps
+		AddPallet::Template => {
+			steps.push(RuntimePalletImport((
+				quote!(
+					// Imports by pop-cli
+					pub use pallet_parachain_template;
+				),
+				2,
+			)));
+			steps.push(SwitchTo(State::Config));
+			steps.push(RuntimePalletConfiguration((
+				quote!(
+					/// Configure the pallet template in pallets/template.
+					impl pallet_parachain_template::Config for Runtime {
+						type RuntimeEvent = RuntimeEvent;
+					}
+				),
+				5,
+			)));
+			steps.push(SwitchTo(State::ConstructRuntime));
+			steps.push(ConstructRuntimeEntry(AddPalletEntry::new(
+				// Index - None, means Pallet Engine will automatically compute an index
+				None,
+				// Path
+				"pallet_parachain_template",
+				// Pallet name
+				// TODO (high priority): implement name conflict resolution strategy
+				"Template",
+			)));
+			steps.push(RuntimePalletDependency(Dependency::local_template_runtime(
+				"pallet-parachain-template",
+				workspace,
+			)));
+		},
+		AddPallet::Frame(_) => unimplemented!("Frame pallets not yet implemented"),
+	};
+	Ok(steps)
+}
+/// Execute steps on PalletEngine.
+/// Each execution edits a file.
+/// Sequence of steps matters so take care when ordering them
+/// Works only for Template pallets at the moment.. See config and CRT inserts
+pub(super) fn run_steps(mut pe: PalletEngine, mut te: TomlEditor, steps: Vec<Steps>) -> Result<()> {
+	use super::State::*;
```

Comment:
States themselves encode the sequence. I have a comment in the State definition that lists a todo to enable sequence checking via discriminants, but I'll leave that out for now, and possibly the future as I feel it would overcomplicate readability.